core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypeConstraints.java

```diff
@@ -0,0 +1,152 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.typereg;
+
+import groovy.xml.Entity;
+
+import java.util.Set;
+
+import org.apache.brooklyn.api.entity.EntitySpec;
+import org.apache.brooklyn.api.internal.AbstractBrooklynObjectSpec;
+import org.apache.brooklyn.api.objs.BrooklynObject;
+import org.apache.brooklyn.api.objs.BrooklynObjectType;
+import org.apache.brooklyn.api.typereg.BrooklynTypeRegistry.RegisteredTypeKind;
+import org.apache.brooklyn.api.typereg.RegisteredTypeConstraint;
+import org.apache.brooklyn.util.collections.MutableSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
+
+public class RegisteredTypeConstraints {
+
+    private static final Logger log = LoggerFactory.getLogger(RegisteredTypeConstraints.BasicRegisteredTypeConstraint.class);
+    
+    public final static class BasicRegisteredTypeConstraint implements RegisteredTypeConstraint {
+        private RegisteredTypeKind kind;
+        private Class<?> javaSuperType;
+        private Set<String> encounteredTypes;
+        
+        private BasicRegisteredTypeConstraint() {}
+        
+        public BasicRegisteredTypeConstraint(RegisteredTypeConstraint source) {
+            if (source==null) return;
+            
+            this.kind = source.getKind();
+            this.javaSuperType = source.getJavaSuperType();
+            this.encounteredTypes = source.getEncounteredTypes();
+        }
+
+        @Override
+        public RegisteredTypeKind getKind() {
+            return kind;
+        }
+        
+        @Override
+        public Class<?> getJavaSuperType() {
+            if (javaSuperType==null) return Object.class;
+            return javaSuperType;
+        }
+
+        @Override
+        public Set<String> getEncounteredTypes() {
+            if (encounteredTypes==null) return ImmutableSet.of();
+            return ImmutableSet.<String>copyOf(encounteredTypes);
+        }
+        
+        @Override
+        public String toString() {
+            return super.toString()+"["+kind+","+javaSuperType+","+encounteredTypes+"]";
+        }
+    }
+
+    /** returns a constraint which allows anything */
+    public static RegisteredTypeConstraint any() {
+        return new BasicRegisteredTypeConstraint();
+    }
+
+    public static RegisteredTypeConstraint alreadyVisited(Set<String> encounteredTypeSymbolicNames) {
+        BasicRegisteredTypeConstraint result = new BasicRegisteredTypeConstraint();
+        result.encounteredTypes = encounteredTypeSymbolicNames;
+        return result;
+    }
+    public static RegisteredTypeConstraint alreadyVisited(Set<String> encounteredTypeSymbolicNames, String anotherEncounteredType) {
+        BasicRegisteredTypeConstraint result = new BasicRegisteredTypeConstraint();
+        result.encounteredTypes = MutableSet.copyOf(encounteredTypeSymbolicNames);
```

Comment:
Better use `ImmutableSet` to avoid accidental changes. Might as well use it for the above method as well, should be a NOP in the general case where we are passing immutable already.